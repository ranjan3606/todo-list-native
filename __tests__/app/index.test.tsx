import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { Text } from 'react-native';
import HomeScreen from '../../app/(tabs)/index';

// Mock required modules
jest.mock('@expo/vector-icons', () => ({
  FontAwesome: 'FontAwesome',
}));

jest.mock('@/constants/Colors', () => ({
  Colors: {
    light: { 
      background: '#fff', 
      text: '#000', 
      tint: '#007AFF',
      lightText: '#666',
      cardBackground: '#f0f0f0'
    },
    dark: { 
      background: '#000', 
      text: '#fff', 
      tint: '#0A84FF',
      lightText: '#999',
      cardBackground: '#1c1c1e'
    },
    primary: '#007AFF',
    danger: '#ff3b30',
  }
}));

jest.mock('@/services/theme', () => ({
  useActualColorScheme: () => 'light',
}));

jest.mock('@/hooks/useTodos', () => ({
  useTodos: () => ({
    incompleteTodos: [],
    isLoading: false,
    refresh: jest.fn(),
  }),
}));

jest.mock('@/utils/dateUtils', () => ({
  categorizeTasks: () => ({
    today: [],
    tomorrow: [],
    upcoming: [],
    past: [],
  }),
}));

// Create simple mocks first
jest.mock('@/components/TaskCard', () => ({
  TaskCard: jest.fn(),
}));

jest.mock('@/components/TaskForm', () => ({
  TaskForm: jest.fn(),
}));

jest.mock('@/i18n', () => ({
  useTranslation: () => ({
    t: (key) => {
      const translations = {
        'tabs.home': 'Home',
        'common.loading': 'Loading',
        'task.today': 'Today',
        'task.tomorrow': 'Tomorrow',
        'task.upcoming': 'Upcoming',
        'task.noTasksToday': 'No tasks for today',
        'task.noTasksTomorrow': 'No tasks for tomorrow',
        'task.noUpcomingTasks': 'No upcoming tasks',
      };
      return translations[key] || key;
    }
  })
}));

jest.mock('@/hooks/useStableRefresh', () => ({
  useStableRefresh: jest.fn(),
}));

jest.mock('@/components/PageLayout', () => ({
  PageLayout: jest.fn(),
}));

// Implement the mock behaviors after the mocks are created
const { TaskCard } = require('@/components/TaskCard');
TaskCard.mockImplementation(({ item }) => (
  <React.Fragment>Task: {item.title}</React.Fragment>
));

const { TaskForm } = require('@/components/TaskForm');
TaskForm.mockImplementation(() => null);

const { PageLayout } = require('@/components/PageLayout');
PageLayout.mockImplementation(({ children, title }) => (
  <React.Fragment>
    <Text>{`HEADER: ${title}`}</Text>
    {children}
  </React.Fragment>
));

describe('HomeScreen', () => {
  it('renders correctly with empty task lists', () => {
    const { getByText, queryByText } = render(<HomeScreen />);
    
    expect(getByText('HEADER: Home')).toBeTruthy();
    expect(getByText('Today')).toBeTruthy();
    expect(getByText('Tomorrow')).toBeTruthy();
    expect(getByText('Upcoming')).toBeTruthy();
    
    // Initially, only "Today" section should be expanded
    expect(getByText('No tasks for today')).toBeTruthy();
    expect(queryByText('No tasks for tomorrow')).toBeNull();
    expect(queryByText('No upcoming tasks')).toBeNull();
  });
  
  it('toggles between sections when headers are clicked', () => {
    const { getByText, queryByText } = render(<HomeScreen />);
    
    // Initially, Today section is active
    expect(getByText('No tasks for today')).toBeTruthy();
    
    // Click on Tomorrow section
    fireEvent.press(getByText('Tomorrow'));
    
    // Now Tomorrow section should be active
    expect(queryByText('No tasks for today')).toBeNull();
    expect(getByText('No tasks for tomorrow')).toBeTruthy();
    
    // Click on Upcoming section
    fireEvent.press(getByText('Upcoming'));
    
    // Now Upcoming section should be active
    expect(queryByText('No tasks for tomorrow')).toBeNull();
    expect(getByText('No upcoming tasks')).toBeTruthy();
  });
  
  it('matches snapshot', () => {
    const { toJSON } = render(<HomeScreen />);
    expect(toJSON()).toMatchSnapshot();
  });
});