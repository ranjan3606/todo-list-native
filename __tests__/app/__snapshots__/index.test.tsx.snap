// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen matches snapshot 1`] = `
[
  <Text>
    HEADER: Home
  </Text>,
  <RCTScrollView
    contentContainerStyle={
      {
        "paddingBottom": 60,
      }
    }
    style={
      {
        "flex": 1,
        "width": "100%",
      }
    }
  >
    <View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#007AFF15",
            "borderLeftColor": "#007AFF",
            "borderLeftWidth": 3,
            "borderRadius": 6,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginVertical": 8,
            "opacity": 1,
            "paddingHorizontal": 10,
            "paddingVertical": 12,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              [
                {
                  "fontSize": 16,
                  "fontWeight": "600",
                },
                {
                  "color": "#000",
                },
                {
                  "color": "#007AFF",
                },
                false,
              ]
            }
          >
            Today
          </Text>
        </View>
        <FontAwesome
          color="#007AFF"
          name="chevron-down"
          size={16}
        />
      </View>
      <Text
        style={
          [
            {
              "fontStyle": "italic",
              "marginTop": 50,
              "marginVertical": 15,
              "textAlign": "center",
            },
            {
              "color": "#666",
            },
          ]
        }
      >
        No tasks for today
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderRadius": 6,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginVertical": 8,
            "opacity": 1,
            "paddingHorizontal": 10,
            "paddingVertical": 12,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              [
                {
                  "fontSize": 16,
                  "fontWeight": "600",
                },
                {
                  "color": "#000",
                },
                false,
                false,
              ]
            }
          >
            Tomorrow
          </Text>
        </View>
        <FontAwesome
          color="#000"
          name="chevron-right"
          size={16}
        />
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderRadius": 6,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginVertical": 8,
            "opacity": 1,
            "paddingHorizontal": 10,
            "paddingVertical": 12,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              [
                {
                  "fontSize": 16,
                  "fontWeight": "600",
                },
                {
                  "color": "#000",
                },
                false,
                false,
              ]
            }
          >
            Upcoming
          </Text>
        </View>
        <FontAwesome
          color="#000"
          name="chevron-right"
          size={16}
        />
      </View>
    </View>
  </RCTScrollView>,
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#007AFF",
        "borderRadius": 28,
        "bottom": 80,
        "elevation": 8,
        "height": 56,
        "justifyContent": "center",
        "opacity": 1,
        "position": "absolute",
        "right": 20,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 3.84,
        "width": 56,
      }
    }
  >
    <FontAwesome
      color="white"
      name="plus"
      size={24}
    />
  </View>,
]
`;
